### PR TITLE
Add isolatedModules error for ambiguous imports referenced in decorator metadata

### LINKED ISSUE #42281: TS is not throughing an error for imports without import type with isolatedModules true
# Bug Report

Typescript is not throughing an error for Type imports not having the keyword `import type` from Files containing only type exports, when having `isolatedModules` set to `true`.

### 🔎 Search Terms

- isolatedModules
- import type

### 🕗 Version & Regression Information

- This is the behaviour in every recent version I tried.

### ⏯ Playground Link

See attached Zip Code and Readme.md for instructions.

### 💻 Code

The types.ts files contains only named exports
```
// types.ts

export interface IGEvent {
	data: {
		eventData: IGEventOptions;
	};
}
```

The index.ts imports the event interface and uses it within it's decorated method signature
```
// index.ts
import {Component, EventListener, GondelBaseComponent} from '@gondel/core';
import { IGEvent } from './types';

@Component('HelloWorld')
class HelloWorld extends GondelBaseComponent {

	@EventListener('gEvent')
	_handleEvent(event: IGEvent) {
		alert(event.data.eventData.helloWorld);
	}
}
```

The used library gondel is a component framework to start components on a page and handle their lifecycle.

The warning goes away when

- we remove the decorator
- we define event as any and cast it inside the method to the type IGEvent
- we import it with import type { IGEvent } from ...

### 🙁 Actual behavior

When using Webpack, it throughs a build-breaking error since Webpack 5, before, a warning was shown. 
When just using native TS for Compiling, I don't get an Error.

### 🙂 Expected behavior

Courtesy of explanation goes to @andrewbranch, see Ticket 40420.

The point of isolatedModules is to tell you ahead of time if a transpiler (like Babel or ts-loader in transpile mode) will not have enough information to know whether the code it produces at runtime is safe or not. That’s what’s happening here—ts-loader in transpile mode cannot possibly know whether IGEvent has a value meaning by only looking at index.ts, and so it stays in. Webpack is smart enough to detect this and tell you about it, and prior to v5, forgiving enough to fix up your runtime code. But at the core, what happened was you wrote something that could not be analyzed by ts-loader, and consequently ts-loader produced fundamentally broken output. isolatedModules is the way you tell the compiler that you’re going to use a transpiler, and so you want errors that prevent you from getting in this situation.
[Code Example.zip](https://github.com/microsoft/TypeScript/files/5794973/Code.Example.zip)

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes #42281 

Teeeechnically maybe a breaking change since there’s a new error, but it should only be an error when a transpiler would produce code that errors at runtime.

### CODE REVIEW

**Comment by rbuckton** [comment]
If I understand correctly, the issue is that we keep in a static import for the interface:

```ts
import { X, B } from "./other";

class C extends B {
  @dec x: X;
}
```

We try to emit `X` as a value using a "fallback" expression:

```js
import { X, B } from "./other";
class C extends B {
}
__decorate([
    dec,
    __metadata("design:type", typeof (_a = typeof X !== "undefined" && X) === "function" ? _a : Object)
], C.prototype, "x", void 0);
```

However, that leaves in a static import for `X`, which may not actually be exported from `"./other"`. What we should do is replace the static import binding with a namespace import:

```js
import { B } from "./other";
import * as other_1 from "./other"; // replace `import { X }` with namespace import...
class C extends B {
}
__decorate([
    dec,
    __metadata("design:type", typeof (_a = typeof other_1.X !== "undefined" && other_1.X) === "function" ? _a : Object)
], C.prototype, "x", void 0);
```

**Comment by rbuckton** [comment]
This doesn't occur in CJS, AMD, or System emit because we always end up with a "namespace" import in those cases:
```js
const other_1 = require("./other");
class C {
}
__decorate([
    dec,
    __metadata("design:type", typeof (_a = typeof other_1.X !== "undefined" && other_1.X) === "function" ? _a : Object)
], C.prototype, "x", void 0);
```

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -879,6 +879,10 @@
         "category": "Error",
         "code": 1266
     },
+    "A type referenced in a decorated signature must be imported with 'import type' when 'isolatedModules' and 'emitDecoratorMetadata' are enabled.": {
```
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/compiler/diagnosticMessages.json:882]: Will add if we decide to keep this approach after discussing in a design meeting

**Comment by andrewbranch** [comment]
I’m not sure where the design meeting notes are where we discussed this, but I believe the compromise we came up with was

- only error in es2015+ (done)
- add codefixes that suggest both `import type` _and_ `import * as whatever` as workarounds

File: src/compiler/diagnosticMessages.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewbranch) [reply src/compiler/diagnosticMessages.json]: Added, along with a fix to convert named imports a namespace import, if it makes sense.

**Comment by andrewbranch** [comment]
I think to be safe, this shouldn’t go in post-beta, so will plan to target for 4.5.

**Review by sandersn — APPROVED** [review]
Looks good to me. I think now is a good time to put it into 4.7 too.

**Pull request merged**